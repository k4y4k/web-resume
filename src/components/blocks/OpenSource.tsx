import dayjs from "dayjs";
import { graphql, useStaticQuery } from "gatsby";
import React from "react";
import ReactMarkdown from "react-markdown";
import ItemTitle from "../section/ItemTitle";
import SectionContainer from "../section/SectionContainer";

interface OpenSource {
  forge: string;
  userRepo: string;
  description: string;
  rawDates?: string;
  relevant?: boolean | null;
}

export const PureOpenSource = ({
  openSource,
}: {
  openSource: OpenSource[];
}) => {
  if (openSource.length === 0) return null;

  return (
    <div data-testid="open-source">
      <SectionContainer title="Open Source">
        {openSource.map(({ forge, userRepo, description, rawDates }) => {
          return (
            <div
              data-testid="sectionItemDetails"
              key={JSON.stringify({ userRepo })}
            >
              <ItemTitle title={userRepo} />

              <div className="italic mb-1">
                <span>
                  {`https://${forge}.com/${userRepo}`}
                  {`${rawDates ? ` | ${rawDates}` : ""}`}
                </span>
              </div>

              <ReactMarkdown>{description}</ReactMarkdown>
            </div>
          );
        })}
      </SectionContainer>
    </div>
  );
};

export const OpenSource = () => {
  const data = useStaticQuery(graphql`
    {
      file(name: { eq: "data" }, extension: { eq: "json" }) {
        id
        childDataJson {
          openSource {
            forge
            userRepo
            description
            relevant
            rawDates
          }
        }
      }
    }
  `);

  const { openSource }: { openSource: OpenSource[] } = data.file.childDataJson;

  const aggregated = [...openSource].filter((el) => {
    // `relevant` is null if unset, but can be set to `true`
    // to exclude an entry, pass in false
    return el.relevant !== false;
  });

  return <PureOpenSource openSource={aggregated} />;
};

export default OpenSource;
