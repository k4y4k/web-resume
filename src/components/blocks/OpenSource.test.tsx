import { render, screen } from "@testing-library/react";
import { useStaticQuery } from "gatsby";
import React from "react";
import type { DataOpenSourceItem } from "../../types/ResumeData";
import OpenSource, { PureOpenSource } from "./OpenSource";

describe("<PureOpenSource />", () => {
  const fixture: DataOpenSourceItem[] = [
    {
      forge: "github",
      description: "- did stuff\n- next dot point",
      userRepo: "username/repository",
    },
  ];

  beforeAll(() => {
    return (useStaticQuery as jest.Mock).mockReturnValue({
      file: {
        childDataJson: {
          openSource: [...fixture],
        },
      },
    });
  });

  test("bails out on no data", () => {
    render(<PureOpenSource openSource={[]} />);

    // still want the title
    expect(screen.getByTestId("open-source")).toBeInTheDocument();

    // if there's no data, we shouldn't see <ItemContainer /> render
    expect(
      screen.queryByTestId("sectionItemContainer"),
    ).not.toBeInTheDocument();
    expect(screen.getByTestId("open-source")).toMatchSnapshot();
  });

  test("renders array of >=1 OpenSource items", () => {
    const { container } = render(<PureOpenSource openSource={[...fixture]} />);

    const openSource = screen.getByTestId("open-source");

    expect(openSource).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders OK", () => {
    render(<OpenSource />);

    expect(screen.getByTestId("open-source")).toMatchSnapshot();
  });
});
