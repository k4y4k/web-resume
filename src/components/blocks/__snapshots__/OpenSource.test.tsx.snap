// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PureOpenSource /> bails out on no data 1`] = `null`;

exports[`<PureOpenSource /> renders OK 1`] = `
<div
  data-testid="open-source"
>
  <div
    class="mx-4 mb-4"
    data-testid="sectionContainer"
  >
    <h2
      class="font-mono text-lg font-bold uppercase text-orchid-600"
      data-testid="sectionTitle"
    >
      Open Source
    </h2>
    <div
      data-testid="sectionItemDetails"
    >
      <h1
        class="mb-1 font-bold text-lg leading-5"
        data-testid="itemTitle"
      >
        username/repository
      </h1>
      <div
        class="italic mb-1"
      >
        <span>
          https://github.com/username/repository
        </span>
      </div>
      <ul>
        

        <li>
          did stuff
        </li>
        

        <li>
          next dot point
        </li>
        

      </ul>
    </div>
  </div>
</div>
`;

exports[`<PureOpenSource /> renders array of >=1 OpenSource items 1`] = `
<div
  data-testid="open-source"
>
  <div
    class="mx-4 mb-4"
    data-testid="sectionContainer"
  >
    <h2
      class="font-mono text-lg font-bold uppercase text-orchid-600"
      data-testid="sectionTitle"
    >
      Open Source
    </h2>
    <div
      data-testid="sectionItemDetails"
    >
      <h1
        class="mb-1 font-bold text-lg leading-5"
        data-testid="itemTitle"
      >
        username/repository
      </h1>
      <div
        class="italic mb-1"
      >
        <span>
          https://github.com/username/repository
        </span>
      </div>
      <ul>
        

        <li>
          did stuff
        </li>
        

        <li>
          next dot point
        </li>
        

      </ul>
    </div>
  </div>
</div>
`;
